core/src/main/java/org/apache/iceberg/view/ViewVersionReplace.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.Tasks;
+
+class ViewVersionReplace implements ReplaceViewVersion {
+  private final ViewOperations ops;
+  private final List<ViewRepresentation> viewRepresentationsToAdd = Lists.newArrayList();
+  private ViewMetadata base;
+  private Namespace defaultNamespace;
+  private String defaultCatalog;
+  private Schema schema;
+
+  ViewVersionReplace(ViewOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+  }
+
+  @Override
+  public ViewVersion apply() {
+    Preconditions.checkState(
+        !viewRepresentationsToAdd.isEmpty(),
+        "Cannot replace view without specifying a dialect/sql");
+
+    this.base = ops.refresh();
+
+    ViewVersion viewVersion = base.currentVersion();
+
+    return ImmutableViewVersion.builder()
+        .versionId(viewVersion.versionId() + 1)
+        .timestampMillis(System.currentTimeMillis())
+        .schemaId(null == schema ? viewVersion.schemaId() : schema.schemaId())
+        .defaultNamespace(defaultNamespace)
+        .defaultCatalog(defaultCatalog)
+        .putSummary("operation", "replace")
+        .addAllRepresentations(viewRepresentationsToAdd)
+        .build();
+  }
+
+  @Override
+  public void commit() {
+    Tasks.foreach(ops)
+        .retry(
+            PropertyUtil.propertyAsInt(
+                base.properties(), COMMIT_NUM_RETRIES, COMMIT_NUM_RETRIES_DEFAULT))
+        .exponentialBackoff(
+            PropertyUtil.propertyAsInt(
+                base.properties(), COMMIT_MIN_RETRY_WAIT_MS, COMMIT_MIN_RETRY_WAIT_MS_DEFAULT),
+            PropertyUtil.propertyAsInt(
+                base.properties(), COMMIT_MAX_RETRY_WAIT_MS, COMMIT_MAX_RETRY_WAIT_MS_DEFAULT),
+            PropertyUtil.propertyAsInt(
+                base.properties(), COMMIT_TOTAL_RETRY_TIME_MS, COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT),
+            2.0 /* exponential */)
+        .onlyRetryOn(CommitFailedException.class)
+        .run(
+            taskOps -> {
+              ViewVersion newVersion = apply();
+              // nothing to do if the version didn't change
+              if (this.base.currentVersion().equals(newVersion)) {
+                return;
+              }
+
+              ViewMetadata updated =
+                  ViewMetadata.buildFrom(this.base)
+                      .setCurrentVersion(
+                          newVersion,
+                          null == schema ? base.schemasById().get(newVersion.schemaId()) : schema)
+                      .build();
+
+              taskOps.commit(base, updated);
+            });
+  }
+
+  @Override
+  public ReplaceViewVersion withQuery(String dialect, String sql) {
+    viewRepresentationsToAdd.add(
+        ImmutableSQLViewRepresentation.builder().dialect(dialect).sql(sql).build());
```

Comment:
Wouldn't we need logic to prevent adding multiple representations of the same dialect? I think for a view version there should only be 1 SQL representation for a given dialect. Also sorry if this was discussed already, I'm late on following up on this

Reply:
I believe we discussed this (in one of the previous PRs or offline), but unfortunately I don't recall what we decided. In any case, I think we can handle this outside the scope of this PR (most likely in `ViewVersion` itself)

Reply:
I think we should add validation here. Putting the logic in `ViewVersion` would cause this to fail when reading a version with duplicate dialects. This should fail to write that metadata, but it should not fail to read.

Reply:
Why do this in a follow-up? There's no reason not to get it done now.

Reply:
the diff for this might be non-trivial (because we need to update this here and potentially also in other places where a view version can be constructed). Hence it seemed for me to make more sense to handle this separately. I was thinking to open a separate PR for this (while this one is being reviewed) and we can then decide to either include the diff in this PR or keep it separate.

Reply:
I've handled this in https://github.com/apache/iceberg/pull/8639 (https://github.com/apache/iceberg/pull/8639/commits/c27bdef65d02300673e2938630f3e3c2f6e9c897). @rdblue let me know if you want me to include the diff as part of #7880